File: Practica 4/CivitasJava4/CivitasJava/src/civitasjava/SorpresaPorJugador.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package civitasjava;

import java.util.ArrayList;

/**
 *
 * @author francisco
 */
public class SorpresaPorJugador extends Sorpresa {
    
    private int valor; // Guarda cuanto se paga por cada jugador
    
    SorpresaPorJugador(int valor, String texto){
        super(texto);
        
        this.valor = valor;
    }
     
    /**
     * @brief Todos los jugadores dan el valor de la sorpresa al jugador actual
     * @param actual, el indice del jugador
     * @param todos, el vector de jugadores
     * @return void
     */
    @Override
    void aplicarAJugador(int actual, ArrayList<Jugador> todos){
      
      if( jugadorCorrecto(actual, todos) ){
        
        this.informe(actual, todos);
        int numJugadores = todos.size();
        
        // Creamos dos Sorpresas con los valores de saldo a modificar
        SorpresaPagarCobrar restoPaga    = new 
                        SorpresaPagarCobrar(-this.valor , "");
        
        SorpresaPagarCobrar actualRecibe = new
                        SorpresaPagarCobrar(this.valor * (numJugadores -1), "");
        
        // Cobramos a todos los jugadores menos el que recibe
        for(int i = 0; i < numJugadores; i++){
            if(i != actual)
                restoPaga.aplicarAJugador(i, todos);    
        }
        
        // Ingresamos el dinero al jugador actual
        actualRecibe.aplicarAJugador(actual, todos);
      }
    }
}

File: Practica 4/CivitasRuby4/CivitasRuby/pkg/RubyApplication1-0.0.1/lib/sorpresa_por_jugador.rb
module Civitas

class Sorpresa_por_jugador < Sorpresa
  def initialize(valor, texto)
    super(texto)
    
    @valor = valor
  end

# @brief Llama al método pagar y cobrar y todos los jugadores le dan @valor
#   al jugador actual, que recibe @valor * num_jugadores 
# @param actual, el indice del jugador
# @param todos, el vector de jugadores
# @return Es void
def aplicar_a_jugador(actual,todos)
  if(jugador_correcto(actual,todos))
    informe(actual,todos)
    
    a_cobrar = Sorpresa_pagar_cobrar.new((todos.length-1) * @valor, "cobrar")
    a_pagar  = Sorpresa_pagar_cobrar.new(              -1 * @valor, "pagar" )
    
    for i in (0..todos.length)
      if(i != actual) #El jugador "actual" no paga
        a_pagar.aplicar_a_jugador(i,todos)
      end
    end
   
    a_cobrar.aplicar_a_jugador(actual,todos) #Llama a modifica saldo 
    
  end
  
end

end

end